sdk/eventhubs/azure_messaging_eventhubs/src/event_processor/processor.rs

```diff
@@ -0,0 +1,637 @@
+    sync::{
+        Arc,
+        Mutex as SyncMutex, // Mutex for blocking operations
+    },
+    time::Duration,
+    {collections::HashMap, sync::Weak},
+};
+use tracing::{debug, error, info};
+
+/// Represents the event processor responsible for processing events
+/// from Event Hub partitions.
+///
+/// This struct manages the load balancing strategy, checkpoint store,
+/// and consumer client for processing events.
+/// It provides methods for starting the event processor, dispatching
+/// events, and managing partition clients.
+///
+/// The event processor uses a load balancer to distribute the load
+/// across partitions and a checkpoint store to manage checkpoints.
+///
+/// For more information on Event Processors and scenarios in which you would
+/// use an Event Processor, see the [Event Processor documentation](https://learn.microsoft.com/azure/event-hubs/event-processor-balance-partition-load).
+///
+pub struct EventProcessor {
+    checkpoint_store: Arc<dyn CheckpointStore + Send + Sync>,
+    load_balancer: Arc<AsyncMutex<LoadBalancer>>,
+    consumer_client: Arc<ConsumerClient>,
+    next_partition_clients: AsyncMutex<Receiver<Arc<PartitionClient>>>,
+    next_partition_client_sender: Sender<Arc<PartitionClient>>,
+    client_details: ConsumerClientDetails,
+    prefetch: u32,
+    update_interval: Duration,
+    start_positions: StartPositions,
+    is_running: std::sync::Mutex<bool>,
+    eventhub_properties: EventHubProperties,
+}
+
+struct EventProcessorOptions {
+    strategy: ProcessorStrategy,
+    partition_expiration_duration: Duration,
+    update_interval: Duration,
+    start_positions: StartPositions,
+    max_partition_count: Option<usize>,
+    prefetch: u32,
+}
+
+pub(crate) struct ProcessorConsumersMap {
+    consumers: SyncMutex<HashMap<String, Weak<PartitionClient>>>,
+}
+
+impl ProcessorConsumersMap {
+    fn new() -> Self {
+        ProcessorConsumersMap {
+            consumers: SyncMutex::new(HashMap::new()),
+        }
+    }
+
+    /// Adds a partition client to the consumers map.
+    /// If a partition client already exists for the given partition ID,
+    /// it will not be added again.
+    /// Returns `true` if the partition client was added successfully,
+    /// or `false` if it already exists.
+    ///
+    /// # Arguments
+    /// * `partition_id` - The ID of the partition for which the client is being added.
+    /// * `partition_client` - The partition client to be added.
+    ///
+    /// # Returns
+    /// A `Result` indicating the success or failure of the operation.
+    /// If successful, returns `true` if the partition client was added,
+    /// or `false` if it already exists.
+    ///
+    pub async fn add_partition_client(
+        &self,
+        partition_id: &str,
+        partition_client: Arc<PartitionClient>,
+    ) -> Result<bool> {
+        info!("Adding partition client for partition: {}", partition_id);
+        let mut consumers = self.consumers.lock().map_err(|_| {
```

Comment:
Why a blocking lock in an async function? Sure, the caller will likely swap context, but seems odd to mix them when you're already using async mutexes elsewhere.

Reply:
Because the lock will only be held for long enough to set a Boolean value to "true" - the additional overhead of an asynchronous lock seems unnecessary.